[FLINK-3874] Rewrite Kafka JSON Table sink tests

mushketyk opened the pull request:
Turned Kafka JSON Table sink tests into unit tests as discussed here: https://issues.apache.org/jira/browse/FLINK-3874

- [x] General
  - The pull request references the related JIRA issue ("[FLINK-XXX] Jira title text")
  - The pull request addresses only one issue
  - Each commit in the PR has a meaningful commit message (including the JIRA id)

- [x] Documentation
  - Documentation has been added for new functionality
  - Old documentation affected by the pull request has been updated
  - JavaDoc for public methods has been added

- [x] Tests & Build
  - Functionality added by the pull request is covered by tests
  - `mvn clean verify` has been executed successfully locally or a Travis build has passed

fhueske commented on flink-streaming-connectors/flink-connector-kafka-base/src/test/java/org/apache/flink/streaming/connectors/kafka/KafkaTableSinkTestBase.java:
```diff
@@ -17,123 +17,60 @@
-		KafkaTableSink kafkaTableSinkBase = createTableSink();
-
-		kafkaTableSinkBase.emitDataStream(stream);
+		verify(dataStream).addSink(kafkaProducer);
```
This test checks that `addSink` was called with a `FlinkKafkaProducerBase` object. I think it would be good to extend the test to make sure that
- the right producer was used (Kafka 0.8, 0.9)
- the correct topic is fetched
- the properties are correctly passed on
- the correct serialization scheme is used
- the correct partitioner is used

Would that be possible?

fhueske commented:
Hi @mushketyk, thanks for reworking the tests and sorry that the review took a while.
I added a comment. Please let me know if you have questions.

Thanks, Fabian

mushketyk commented:
Hi @fhueske,

Thank you for your review. I'll take a look at the code, check if these checks are possible to implement and come back with a reply.

mushketyk commented:
Hey @fhueske 

I've updated the PR and implemented most of your suggestion. I am not sure how to check if the right producer version is used thought.
The problem is that to check the producer type we need to created its instance, and when an instance is created it immediately tries to connect to Kafka. I can start test Kafka to overcome this, but it seems like an overkill and looks like a step in the opposite direction. 

What do you think about this?

fhueske commented on flink-streaming-connectors/flink-connector-kafka-base/src/test/java/org/apache/flink/streaming/connectors/kafka/KafkaTableSinkTestBase.java:
```diff
@@ -17,123 +17,79 @@
+		return new Properties();
 	}
 
 	protected abstract KafkaTableSink createTableSink();
```
I would add the required parameter for a KafkaJsonTableSink here (topic, props, partitioner). That will make the code a bit easier to understand, IMO.

fhueske commented on flink-streaming-connectors/flink-connector-kafka-base/src/test/java/org/apache/flink/streaming/connectors/kafka/KafkaTableSinkTestBase.java:
```diff
@@ -17,123 +17,79 @@
+		assertEquals(new RowTypeInfo(FIELD_TYPES), newKafkaTableSink.getOutputType());
 	}
 
 	protected KafkaPartitioner<Row> createPartitioner() {
```
Please add a partitioner as final member variable to the test which is used.

fhueske commented on flink-streaming-connectors/flink-connector-kafka-base/src/test/java/org/apache/flink/streaming/connectors/kafka/KafkaTableSinkTestBase.java:
```diff
@@ -17,123 +17,79 @@
 		return new CustomPartitioner();
 	}
 
 	protected Properties createSinkProperties() {
```
Please add a Properties object (non-empty) as a final member variable to the test base which is used. This will make the equals check more meaningful.

mushketyk commented:
Hi @fhueske 
Thank you for your review. I've updated my PR accordingly.

twalthr commented:
@mushketyk thanks. I will merge this.
